File: src/Git.Domain/Paging.cs
using System;
using System.Collections.Generic;
using System.Linq;
using Newtonsoft.Json;

namespace Git.Domain
{
    public class Paging<T>
    {
        private const int MinimumPageSize = 1;

        public Paging(long total, IEnumerable<T> data, int page, int pageSize, int lastPage)
        {
            Total = total;
            Data = data;
            Page = page;
            PageSize = pageSize;
            LastPage = lastPage;
        }

        [JsonProperty(PropertyName = "data")]
        public IEnumerable<T> Data { get; }
        [JsonProperty(PropertyName = "page")]
        public int Page { get; }      
        [JsonProperty(PropertyName = "nextPage")]
        public int? NextPage
        {
            get
            {
                if (Page < LastPage)
                {
                    return Page + 1;
                }
                return null;
            }
        }
        [JsonProperty(PropertyName = "previousPage")]
        public int? PreviousPage
        {
            get
            {
                if (Page == 1)
                {
                    return null;                    
                }

                return Page - 1;
            }
        }
        [JsonProperty(PropertyName = "lastPage")]
        public int LastPage { get; }
        [JsonProperty(PropertyName = "pageSize")]
        public int PageSize { get; }
        [JsonProperty(PropertyName = "total")]
        public long Total { get; }

        public static Paging<T> Create(long total, IEnumerable<T> data, int page, int pageSize, int lastPage)
        {
            return new Paging<T>(total, data, page, pageSize, lastPage);
        }

        public static Paging<T> Generate(IEnumerable<T> allData, int pageSize, int page)
        {
            if (allData == null)
            {
                throw new ArgumentNullException(nameof(allData));
            }

            if (pageSize <= 0)
            {
                pageSize = MinimumPageSize;
            }

            int total = allData.Count();
            var maxPageCount = total % pageSize != 0
                ? total / pageSize + 1
                : total / pageSize;

            var zeroIndexedCurrentPage = page - 1;
            if (IsCurrentPageLessThanZero(zeroIndexedCurrentPage))
            {
                zeroIndexedCurrentPage = 0;
            }

            if (IsPageGreaterThanZeroOrLastPage(page, maxPageCount))
            {
                zeroIndexedCurrentPage = maxPageCount - 1;
            }

            var skip = zeroIndexedCurrentPage * pageSize;
            var data = allData.Skip(skip).Take(pageSize);
            return Paging<T>.Create(total, data, zeroIndexedCurrentPage + 1, data.Count(), maxPageCount);
        }

        private static bool IsCurrentPageLessThanZero(int zeroIndexedCurrentPage)
        {
            return zeroIndexedCurrentPage < 0;
        }

        private static bool IsPageGreaterThanZeroOrLastPage(int page, int maxPageCount)
        {
            return maxPageCount > 0 && page >= maxPageCount;
        }
    }
}

File: src/node-accident-statistics/src/models/paging.ts
import { ExtendedArray } from "./../arrays/extendedArray";

export class Paging<T> {

    public data?: T[];
    public lastPage?: number;
    public nextPage?: number;
    public page?: number;
    public previousPage?: number;
    public pageSize?: number;
    public total?: number;

    public generate(allData: ExtendedArray<T>, page: number, pageSize: number): Paging<T> {
        if (pageSize <= 0) {
            throw new Error("Page size can not be less than or equal to zero");
        }
        const total = allData.length;
        const maxPageCount = Math.ceil(total / pageSize);
        let zeroIndexedCurrentPage = page - 1;

        if (zeroIndexedCurrentPage < 0) {
            zeroIndexedCurrentPage = 0;
        }

        if (maxPageCount > 0 && page >= maxPageCount) {
            zeroIndexedCurrentPage = maxPageCount - 1;
        }

        const skip = zeroIndexedCurrentPage * pageSize;
        const data = allData.skip(skip).take(pageSize);
        const result = this.create(data, zeroIndexedCurrentPage + 1, total, maxPageCount);
        return result;
    }

    private create(data: ExtendedArray<T>, page: number, total: number, maxPageCount: number): Paging<T> {
        const result = new Paging<T>();
        result.data = data;
        result.page = page;
        result.total = total;
        result.pageSize = data.length;
        result.lastPage = maxPageCount;
        if (result.page && result.lastPage && result.page < result.lastPage) {
            result.nextPage = result.page + 1;
        }
        if (result.page && result.page > 1) {
            result.previousPage = result.page - 1;
        }
        return result;
    }
}
